### PR TITLE
proselint 0.7.0 (new formula)

### PR DESCRIPTION
**Pull request by habibalamin**

### CODE REVIEW

**Comment by ilovezfs** [comment]
did you try using virtualenv?

**Comment by habibalamin** [comment]
No. I'm not sure why virtualenv would be useful here. `proselint` is a program meant to be used by an end-user, not a library meant to be used by a developer.

**Comment by ilovezfs** [comment]
Right. I mean the virtualenv feature. See certbot for an example.

**Comment by ilovezfs** [comment]
You can probably get away with just `virtualenv_install_with_resources` here. See alot aws-elasticbeanstalk charm-tools cheat cookiecutter csvtomd dbt duplicity dxpy euca2ools fdroidserver flake8 fonttools gitless httpie internetarchive jrnl khard legit logentries mdv mimms mkdocs molecule mongo-orchestration mps-youtube mycli nodeenv ooniprobe passpie pius platformio pre-commit pyinvoke remarshal rtv saltstack scour shyaml sqlparse sshuttle statik stormpath-cli subliminal supervisor svtplay-dl tarsnapper termius thefuck todoman translate-toolkit twarc vdirsyncer watson xonsh you-get

**Comment by habibalamin** [comment]
Ah, thanks. That clarifies things. I thought you were talking about Let's Encrypt's use of `virtualenv`/`docker`(?) in their certificate creation/deployment process. A search for `certbot virtualenv` didn't return anything helpful.

You're talking about the formula for `certbot`. I'll take a look. The way I did it was inspired by older formulae.

**Comment by ilovezfs** [comment]
Yeah certbot contains a complex example for your future reference, but this formula is probably lucky and can use the simpler virtualenv_install_with_resources DSL.

File: Formula/proselint.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+class Proselint < Formula
+  include Language::Python::Virtualenv
+
+  desc "Linter for prose"
+  homepage "http://proselint.com"
+  url "https://pypi.python.org/packages/e6/78/c46f694a0e43ce47d0d6ae089f750131544018dc1119fc7da58ffd4b1e03/proselint-0.7.0.tar.gz"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: https://files.pythonhosted.org/packages/e6/78/c46f694a0e43ce47d0d6ae089f750131544018dc1119fc7da58ffd4b1e03/proselint-0.7.0.tar.gz

**Comment by habibalamin** [comment]
By the way, I spent some time looking for a way to install Python (well, `pip`) programs and didn't see anything. Is there some part of the documentation that I missed?

I love that you guys recognise that some programs belong in a general purpose system-level package manager, rather than a language project-level package manager, like PyPI, regardless of the language they're written in. I wish the authors themselves often realised this.

This is a big pet peeve of mine. It's especially annoying, because the use of version managers often interferes with the installation of that kind of software.

File: Formula/proselint.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  depends_on :python if MacOS.version <= :snow_leopard
+
+  resource "click" do
+    url "https://pypi.python.org/packages/95/d9/c3336b6b5711c3ab9d1d3a80f1a3e2afeb9d8c02a7166462f6cc96570897/click-6.7.tar.gz"
+    sha256 "f15516df478d5a56180fbf80e68f206010e6d160fc39fa508b65e035fd75130b"
+  end
+
+  resource "future" do
+    url "https://pypi.python.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: https://files.pythonhosted.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz

File: Formula/proselint.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  desc "Linter for prose"
+  homepage "http://proselint.com"
+  url "https://pypi.python.org/packages/e6/78/c46f694a0e43ce47d0d6ae089f750131544018dc1119fc7da58ffd4b1e03/proselint-0.7.0.tar.gz"
+  sha256 "094d808d44bf1a60dcb1465749be5cc44f4f6c146c04bc5f28976a833786e830"
+
+  depends_on :python if MacOS.version <= :snow_leopard
+
+  resource "click" do
+    url "https://pypi.python.org/packages/95/d9/c3336b6b5711c3ab9d1d3a80f1a3e2afeb9d8c02a7166462f6cc96570897/click-6.7.tar.gz"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: https://files.pythonhosted.org/packages/95/d9/c3336b6b5711c3ab9d1d3a80f1a3e2afeb9d8c02a7166462f6cc96570897/click-6.7.tar.gz

File: Formula/proselint.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    sha256 "f15516df478d5a56180fbf80e68f206010e6d160fc39fa508b65e035fd75130b"
+  end
+
+  resource "future" do
+    url "https://pypi.python.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz"
+    sha256 "e39ced1ab767b5936646cedba8bcce582398233d6a627067d4c6a454c90cfedb"
+  end
+
+  resource "six" do
+    url "https://pypi.python.org/packages/source/s/six/six-1.10.0.tar.gz"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: https://files.pythonhosted.org/packages/b3/b2/238e2590826bfdd113244a40d9d3eb26918bd798fc187e2360a8367068db/six-1.10.0.tar.gz

File: Formula/proselint.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+class Proselint < Formula
+  include Language::Python::Virtualenv
+
+  desc "Linter for prose"
+  homepage "http://proselint.com"
+  url "https://pypi.python.org/packages/e6/78/c46f694a0e43ce47d0d6ae089f750131544018dc1119fc7da58ffd4b1e03/proselint-0.7.0.tar.gz"
+  sha256 "094d808d44bf1a60dcb1465749be5cc44f4f6c146c04bc5f28976a833786e830"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: head "https://github.com/amperser/proselint.git"

File: Formula/proselint.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    url "https://pypi.python.org/packages/source/s/six/six-1.10.0.tar.gz"
+    sha256 "105f8d68616f8248e24bf0e9372ef04d3cc10104f1980f54d57b2ce73a5ad56a"
+  end
+
+  def install
+    virtualenv_install_with_resources
+  end
+
+  test do
+    system "#{bin}/proselint", "--version"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: would prefer a "real" test since it's a new formula

**Comment by ilovezfs** [comment]
@habibalamin thanks. Yes, if you want pip, `brew install python` or if you want pip3, `brew install python3`

**Comment by habibalamin** [comment]
Ah, yes. So this would also depend on `:python`, too, regardless of OS version, because OS X doesn't come with `pip`, does it?

File: Formula/proselint.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
+class Proselint < Formula
+  include Language::Python::Virtualenv
+
+  desc "Linter for prose"
+  homepage "http://proselint.com"
+  url "https://files.pythonhosted.org/packages/e6/78/c46f694a0e43ce47d0d6ae089f750131544018dc1119fc7da58ffd4b1e03/proselint-0.7.0.tar.gz"
+  sha256 "094d808d44bf1a60dcb1465749be5cc44f4f6c146c04bc5f28976a833786e830"
```
Comment (ilovezfs) [inline Formula/proselint.rb]: head "https://github.com/amperser/proselint.git"

Comment (habibalamin) [inline Formula/proselint.rb]: Huh, I thought I added that. Cheers!

**Comment by ilovezfs** [comment]
No we have our own vendored pip for the virtualenvs. I meant if you wanted pip for your own use.

File: Formula/proselint.rb (hunk 7 above)
Comment (ilovezfs) [inline Formula/proselint.rb]: thanks. it's just so livecheck can find the new versions :) alternatively we could use the github tarball for the `url` field but people have a tendency to come along and replace that with the pypi links lol

File: Formula/proselint.rb (hunk 6 above)
Comment (habibalamin) [inline Formula/proselint.rb]:
```
    File.open("prose.txt", "w+") do |file|
      file.write "John is very unique."
    end

    IO.popen("#{bin}/proselint --compact prose.txt") do |process|
      actual_result = process.read

      expected_result = "-:1:9: weasel_words.very Substitute 'damn' every time you're inclined " <<
      "to write 'very;' your editor will delete it and the writing will be just as it should be.\n" <<
      "-:1:9: uncomparables.misc Comparison of an uncomparable: 'very unique.' is not comparable."

      actual_result == expected_result
    end
```

Do you think backticks would be fine for this?

**Comment by habibalamin** [comment]
I do have `pip` for my own use, but I don't think installing system-level programs through project-level package managers works very well. Even if you don't use version managers, you just end up having to add junk to your `PATH`. I shouldn't, as a user, have to use `gem`, `pip`, `cabal`, `cargo`, `go`, or anything like that, *regardless of whether I have it installed or not*.

It's not an acceptable distribution method, except for libraries and project tools; project tools may include binaries, but those are easily used per-project, through techniques like prepending (`bundle exec`), running a single command when you start working on a project (how `virtualenv` does it, I think, it's been a while), and probably more.

File: Formula/proselint.rb (hunk 6 above)
Comment (ilovezfs) [inline Formula/proselint.rb]: I would do this:
```
  test do
    (testpath/"prose.txt").write("John is very unique.\n")
    output = shell_output("#{bin}/proselint --compact prose.txt", 1)
    assert_match /weasel_words.*uncomparables/m, output
  end
```

Comment (ilovezfs) [inline Formula/proselint.rb]: Or even just
```
    output = pipe_output("#{bin}/proselint --compact -", "John is very unique.")
    assert_match /weasel_words.*uncomparables/m, output
```

File: Formula/proselint.rb (hunk 7 above)
Comment (habibalamin) [inline Formula/proselint.rb]: Ah, I actually did use the GitHub tarball originally, but I wanted it to be consistent with the resources, which were all being fetched from PyPI.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Shipped! @habibalamin thanks for the :new: formula! :rocket:

**Pull request closed without merging**